controller_manager:controller_manager:
  ros__parameters:
    update_rate: 1000
    fr5_joint_trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController
    gripper_controller:
      type: position_controllers/GripperActionController
    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

fr5_joint_trajectory_controller:
  ros__parameters:
    joints:
      - joint1
      - joint2
      - joint3
      - joint4
      - joint5
      - joint6
    
    # 【关键修改】这里必须是 effort，才能让 PID 产生扭矩去对抗重力
    command_interfaces:
      - effort
    
    state_interfaces:
      - position
      - velocity
    
    state_publish_rate: 50.0
    action_monitor_rate: 20.0
    
    # PID 参数 (之前你加的这些，现在终于能派上用场了！)
    gains:
      joint1: {p: 20000.0, i: 500.0, d: 200.0, i_clamp: 1000.0}
      joint2: {p: 20000.0, i: 500.0, d: 200.0, i_clamp: 1000.0}
      joint3: {p: 15000.0, i: 300.0, d: 150.0, i_clamp: 500.0}
      joint4: {p: 10000.0, i: 200.0, d: 100.0, i_clamp: 500.0}
      joint5: {p: 5000.0,  i: 100.0, d: 50.0,  i_clamp: 100.0}
      joint6: {p: 2000.0,  i: 50.0,  d: 20.0,  i_clamp: 50.0}

gripper_controller:
  ros__parameters:
    joint: left_finger_joint
    action_monitor_rate: 20.0
    goal_tolerance: 0.02
    max_effort: 500.0
    allow_stalling: true 
    gains:
      left_finger_joint: {p: 2000.0, i: 10.0, d: 20.0, i_clamp: 50.0}